### ISSUE TITLE
Suggestion: Track modified branches during any operation, and only reset those branches during `git undo`

### ISSUE THREAD

**Issue by stephenwade** [git-town/git-town#3765]
### Your message to us

I thought of this after reading [`git town undo` mysteriously force-pushed to a bunch of unrelated feature branches #3763](https://github.com/git-town/git-town/issues/3763).

Git Town should keep track of what branches it actually touches and only undo changes to those branches. #3764 fixes the case where it would do damage to the remote, but I worry that it's still possible for me to make local changes while a Git Town command runs.

I know I've gotten distracted before during a rebase conflict and gone off to make other branches and commits, forgetting I'm in the middle of a rebase in a different terminal window. I'm imagining this scenario:

1. The setting "sync-feature-strategy" is set to "rebase".
2. It's 9 AM. I run `git sync`, and it has a rebase conflict. I don't notice the conflict message.
3. In another terminal, I create a branch for a new feature with `git checkout` and make some commits.
4. It's 1 PM. I notice the conflict in the first terminal. I don't remember what that rebase was doing, so I run `git undo`.
5. Does Git Town delete my new branch??

**Comment by kevgo** [comment]
Good example. Yes, `git town undo` as it works right now would delete your new local branch in this scenario. Not an acceptable outcome. Let's improve it!

Your proposed solution sounds like the right approach. Technically this could be done quite easily:

- Add a method `func BranchNames() []gitdomain.BranchName` to the [Opcode](src/vm/shared/opcode.go) interface. This method provides the names of all branches that an opcode instance touches.
- Create a default implementation for this method in [src/vm/opcodes/core.go](https://github.com/git-town/git-town/blob/main/src/vm/opcodes/core.go). This default implementation uses reflection to return the content of all fields that have type `gitdomain.LocalBranchName`, `gitdomain.RemoteBranchName`, or `gitdomain.BranchName`. This saves us from having to implement and maintain a `BranchNames` implementation for all 61 opcodes.
- Individual opcode types can override this method in case the generic lookup misses something. 
- Add a method `func BranchNames() []gitdomain.BranchName` to [Program](src/vm/program/program.go). This method provides the names of all branches touched by the program (which is just a collection of opcodes).
- Store the list of touched branches in the persisted runstate so that it is available when calculating the undo program.
- When creating the undo program, include only the branches stored in the persisted runstate.

**Comment by stephenwade** [comment]
Hmm, I just tried to reproduce what I thought would happen, and I was surprised. Git Town didn't delete my new branch.

I think I'm missing something in my mental model of how `git undo` works. Based on [this comment](https://github.com/git-town/git-town/issues/3763#issuecomment-2251435076), I expected my new branch to be gone after running `git undo`, but it wasn't.

Here's my terminal history:

```
~/git/test$ git town --version
Git Town 14.3.1
~/git$ mkdir test; cd test; git init
Initialized empty Git repository in /Users/stephen.wade/git/test/.git/
~/git/test$ # Step 0. Set up a rebase conflict
~/git/test$ touch README; git add .; git commit -m 'First commit'
[main (root-commit) 009d221] First commit
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 README
~/git/test$ git hack snazzy-readme
Git Town needs to be configured

Main branch: main

[main] git checkout -b snazzy-readme
Switched to a new branch 'snazzy-readme'

~/git/test$ echo 'This is a really cool project.' > README
~/git/test$ git commit -am 'Snazzy!'
[snazzy-readme 28e27ad] Snazzy!
 1 file changed, 1 insertion(+)
~/git/test$ git switch main
Switched to branch 'main'
~/git/test$ echo 'This is my project.' > README
~/git/test$ git commit -am 'Add README'
[main 461f08a] Add README
 1 file changed, 1 insertion(+)
~/git/test$ # Step 1. The setting "sync-feature-strategy" is set to "rebase".
~/git/test$ git config git-town.sync-feature-strategy rebase
~/git/test$ # Step 2. I run `git sync`, and it has a rebase conflict.
~/git/test$ git switch snazzy-readme; git sync
Switched to branch 'snazzy-readme'

[snazzy-readme] git rebase main
Auto-merging README
CONFLICT (content): Merge conflict in README
error: could not apply 28e27ad... Snazzy!
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
hint: Disable this message with "git config advice.mergeConflict false"
Could not apply 28e27ad... Snazzy!


Error: exit status 1

To continue after having resolved conflicts, run "git town continue".
To go back to where you started, run "git town undo".
To continue by skipping the current branch, run "git town skip".


~/git/test$ # Step 3. I create another branch with `git checkout` and make some commits.
~/git/test$ git checkout -b distracted
Switched to a new branch 'distracted'
~/git/test$ echo 'These are my really important plans' > PLANS
~/git/test$ git add .; git commit -m 'Sure hope something doesn’t happen to this'
[distracted 04d54bf] Sure hope something doesn’t happen to this
 2 files changed, 5 insertions(+)
 create mode 100644 PLANS
~/git/test$ # Step 4. I notice the conflict in the first terminal and run `git undo`.
~/git/test$ git undo
git: 'undo' is not a git command. See 'git --help'.

The most similar command is
	bundle
~/git/test$ git town undo

[distracted] git rebase --abort

[distracted] git checkout snazzy-readme
Already on 'snazzy-readme'

~/git/test$ git branch
  distracted
  main
* snazzy-readme
~/git/test$ git show --quiet distracted
commit 04d54bf1ab7298ef83810f542557431f2c1d3ff4 (distracted)
Author: Stephen Wade <stephen.wade@avb.net>
Date:   Fri Jul 26 11:34:11 2024 -0400

    Sure hope something doesn’t happen to this

**Comment by kevgo** [comment]
One more detail that is important to understand how `git town undo` works: it uses the "after snapshot" from the saved runstate, i.e. the repo status that existed after the last Git Town command ran. This explains the behavior you see:

1. You run `git sync` on your `snazzy-readme` branch. It does these things:
   - capture the "before snapshot"
   - run the sync logic
   - while syncing it runs into a merge conflict
   - capture the "after snapshot"
   - persist the before and after snapshot to disk
   - exit complaining about the merge conflict
2. You create your new `distracted` branch.
3. You run `git town undo`. It does these things:
   - Load the persisted "after snapshot" from disk. Note that this snapshot was taken before you created your `distracted` branch and therefore does not contain it.
   - It correctly undoes only the sync of your `snazzy-readme` branch.

To experience the problematic behavior, you need to create your `distracted` branch and then run `git town continue` so that Git Town creates a new "after snapshot" that contains the new branch.

1. Run `git sync` on the `snazzy-readme` branch and see it fail with the merge conflict.
2. Create the `distracted` branch in another terminal.
3. Go back to your first terminal and resolve the merge conflict.
4. Run `git town continue` on the `snazzy-readme` branch. It does these things:
   - continue the sync logic
   - when done (or running into another merge conflict), it captures a new "after snapshot" and overrides the old "after snapshot" on disk with the new one.
5. Run `git town undo`. Now the persisted after snapshot contains the `distracted` branch and Git Town deletes it.

Git Town has a pretty snazzy end-to-end testing framework that makes it easy to describe and execute such scenarios programmatically but in a nicely human-readable format. Here is a failing end-to-end test that reproduces the scenario we are talking about: [features/sync/current_branch/feature_branch/create_branch_in_between.feature](https://github.com/git-town/git-town/blob/kg-reset-only-touched-branches/features/sync/current_branch/feature_branch/create_branch_in_between.feature).

The section called `Background:` sets up the test environment. The section called `Scenario: undo` displays the problematic behavior: it runs `git branch -D feature-2`. `feature-2` is a branch I created while resolving a merge conflict on branch `feature-1`.

If you have Go 1.22 installed, you can run this end-to-end test via this command:

```
go test -- features/sync/current_branch/feature_branch/create_branch_in_between.feature
```

**Comment by stephenwade** [comment]
Thank you for the thorough explanation!

**Comment by kevgo** [comment]
@stephenwade The approach you have come up with and described in this ticket is brilliant and much more effective than my previous approach to exclude remote-only branches from undo. Thanks for sharing it! 🙏 I have implemented this ticket in https://github.com/git-town/git-town/pull/3767. It works so well that I could remove the code from my previous approach altogether!

**Comment by kevgo** [comment]
This is shipped in [v14.4.1](https://github.com/git-town/git-town/releases/tag/v14.4.1).
